PolyEngine/Engine/Src/Engine.hpp

```diff
@@ -1,31 +1,72 @@
 #pragma once
 
+#include <functional>
 #include <Core.hpp>
 #include "RenderingContext.hpp"
 #include "World.hpp"
+#include "SystemBase.hpp"
 
-namespace Poly {
+namespace Poly 
+{
 	class Engine;
+	typedef std::function<void(World*)> PhaseUpdateFunction;
 
-	class ENGINE_DLLEXPORT IGame : public BaseObject<> {
+	class ENGINE_DLLEXPORT IGame : public BaseObject<> 
+	{
 	public:
 		virtual void RegisterEngine(Engine* engine) = 0;
 		virtual void Init() = 0;
 		virtual void Update(float dt) = 0;
 	};
 
-	class ENGINE_DLLEXPORT Engine : public BaseObject<> {
+	class ENGINE_DLLEXPORT Engine : public BaseObject<> 
+	{
 	public:
 		Engine(IGame* game);
 		bool Init(const IRenderingContextParams* context);
 		void Deinit();
 
+		//////////////////////////////
+		/// Registers a PhaseUpdateFunction to be executed in the Preupdate 
+		/// part of a single frame in the same order as they were passed in.
+		/// @see RegisterUpdatePhase()
+		/// @see RegisterPostupdatePhase()
+		/// @param phaseFunction - void function(World*)
+		//////////////////////////////
+		void RegisterPreupdatePhase(PhaseUpdateFunction phaseFunction);
```

Comment:
We can avoid code duplication (triplication?) by using enum class designating all three phases as an input argument.